xdem/spatial_tools.py

```diff
@@ -398,6 +509,336 @@ def hillshade(dem: Union[np.ndarray, np.ma.masked_array], resolution: Union[floa
+    costs = np.empty(max_order)
+    coeffs = np.zeros((max_order, max_order + 1))
+    # loop on polynomial degrees
+    for deg in np.arange(1, max_order + 1):
+        # if method is linear, and package is scipy
+        if estimator == 'Linear' and linear_pkg == 'scipy':
+            # define the residual function to optimize
+            def fitfun_polynomial(xx, params):
+                return sum([p * (xx ** i) for i, p in enumerate(params)])
+            def errfun(p, xx, yy):
+                return fitfun_polynomial(xx, p) - yy
+            p0 = np.polyfit(x, y, deg)
+            myresults = scipy.optimize.least_squares(errfun, p0, args=(x, y), **kwargs)
+            if verbose:
+                print("Initial Parameters: ", p0)
+                print("Polynomial degree - ", deg, " --> Status: ", myresults.success, " - ", myresults.status)
+                print(myresults.message)
+                print("Lowest cost:", myresults.cost)
+                print("Parameters:", myresults.x)
+            costs[deg - 1] = myresults.cost
+            coeffs[deg - 1, 0:myresults.x.size] = myresults.x
+        # otherwise, it's from sklearn
+        else:
+            if not _has_sklearn:
+                raise ValueError("Optional dependency needed. Install 'scikit-learn'")
+
+            # create polynomial + linear estimator pipeline
+            p = PolynomialFeatures(degree=deg)
+            model = make_pipeline(p, est)
+
+            # TODO: find out how to re-scale polynomial coefficient + doc on what is the best scaling for polynomials
+            # # scale output data (important for ML algorithms):
+            # robust_scaler = RobustScaler().fit(x.reshape(-1,1))
+            # x_scaled = robust_scaler.transform(x.reshape(-1,1))
+            # # fit scaled data
+            # model.fit(x_scaled, y)
+            # y_pred = model.predict(x_scaled)
+
+            # fit scaled data
+            model.fit(x.reshape(-1,1), y)
+            y_pred = model.predict(x.reshape(-1,1))
+
+            # calculate cost
+            cost = cost_func(y_pred, y)
+            costs[deg - 1] = cost
+            # get polynomial estimated with the estimator
+            if estimator in ['Linear','Theil-Sen','Huber']:
+                c = est.coef_
+            # for some reason RANSAC doesn't store coef at the same place
+            elif estimator == 'RANSAC':
+                c = est.estimator_.coef_
+            coeffs[deg - 1, 0:deg+1] = c
+
+    # selecting the minimum (not robust)
+    # final_index = mycost.argmin()
+    # choosing the best polynomial with a margin of improvement on the cost
+    final_index = _choice_best_order(cost=costs, margin_improvement=margin_improvement, verbose=verbose)
+
+    # the degree of the polynom correspond to the index plus one
+    return np.trim_zeros(coeffs[final_index], 'b'), final_index + 1
+
+
+def _fitfun_sumofsin(x: np.array, params: np.ndarray) -> np.ndarray:
```

Comment:
The name is not very adequate. You are not fitting a function, but calculating its values.
Maybe sumofsin_val? (to be somewhat similar to np.polyval)

Reply:
Refactored into `sumofsinval` :)